File: sdk/python/velocity_sdk/update.py
"""
Workflow Update API — synchronous workflow mutation.

Unlike signals (fire-and-forget), updates provide:
- Synchronous request/response semantics
- Wait policies (Accepted, Completed, Admitted)
- Validation before execution
- Named update handlers registered by workflows

Usage:
    from velocity_sdk.update import UpdateClient, UpdateWaitPolicy

    client = UpdateClient("localhost:7234")
    result = client.execute_update(
        workflow_key=42,
        update_name="setAmount",
        args={"amount": 100},
        wait_policy=UpdateWaitPolicy.COMPLETED,
    )
"""

from dataclasses import dataclass, field
from enum import IntEnum
from typing import Any, Callable, Dict, List, Optional
import time
import threading


class UpdateStatus(IntEnum):
    """Status of a workflow update."""
    ADMITTED = 0
    ACCEPTED = 1
    COMPLETED = 2
    REJECTED = 3


class UpdateWaitPolicy(IntEnum):
    """How long to wait for an update to complete."""
    ADMITTED = 0
    ACCEPTED = 1
    COMPLETED = 2


@dataclass
class UpdateRequest:
    """Request to execute a workflow update."""
    workflow_key: int
    update_id: str
    update_name: str
    args: Any = None
    wait_policy: UpdateWaitPolicy = UpdateWaitPolicy.COMPLETED


@dataclass
class UpdateResult:
    """Result of a workflow update execution."""
    update_id: str
    status: UpdateStatus
    result: Any = None
    error: Optional[str] = None
    duration_ms: float = 0.0


@dataclass
class UpdateHandler:
    """Handler for a named update."""
    name: str
    handler: Callable[[Any], Any]
    validator: Optional[Callable[[Any], bool]] = None


class UpdateClient:
    """
    Client for executing workflow updates.

    Updates are synchronous mutations to workflow state, unlike signals
    which are fire-and-forget. Updates can be validated before execution
    and support wait policies for different completion levels.
    """

    def __init__(self, server_address: str = "localhost:7234"):
        self._server_address = server_address
        self._handlers: Dict[str, UpdateHandler] = {}
        self._pending: Dict[str, UpdateResult] = {}
        self._lock = threading.Lock()

    def register_handler(
        self,
        name: str,
        handler: Callable[[Any], Any],
        validator: Optional[Callable[[Any], bool]] = None,
    ) -> None:
        """Register a named update handler."""
        self._handlers[name] = UpdateHandler(name=name, handler=handler, validator=validator)

    def execute_update(
        self,
        workflow_key: int,
        update_name: str,
        args: Any = None,
        wait_policy: UpdateWaitPolicy = UpdateWaitPolicy.COMPLETED,
        update_id: Optional[str] = None,
    ) -> UpdateResult:
        """
        Execute a workflow update.

        Args:
            workflow_key: Target workflow key.
            update_name: Name of the registered update handler.
            args: Arguments to pass to the handler.
            wait_policy: How long to wait for completion.
            update_id: Optional update ID (auto-generated if not provided).

        Returns:
            UpdateResult with status and result/error.
        """
        uid = update_id or f"update-{workflow_key}-{int(time.time() * 1000)}"
        start = time.time()

        handler = self._handlers.get(update_name)
        if handler is None:
            result = UpdateResult(
                update_id=uid,
                status=UpdateStatus.REJECTED,
                error=f"No handler registered for update '{update_name}'",
                duration_ms=(time.time() - start) * 1000,
            )
            with self._lock:
                self._pending[uid] = result
            return result

        # Validate if validator exists
        if handler.validator and not handler.validator(args):
            result = UpdateResult(
                update_id=uid,
                status=UpdateStatus.REJECTED,
                error="Update validation failed",
                duration_ms=(time.time() - start) * 1000,
            )
            with self._lock:
                self._pending[uid] = result
            return result

        # Execute the handler
        try:
            value = handler.handler(args)
            result = UpdateResult(
                update_id=uid,
                status=UpdateStatus.COMPLETED,
                result=value,
                duration_ms=(time.time() - start) * 1000,
            )
        except Exception as e:
            result = UpdateResult(
                update_id=uid,
                status=UpdateStatus.REJECTED,
                error=str(e),
                duration_ms=(time.time() - start) * 1000,
            )

        with self._lock:
            self._pending[uid] = result
        return result

    def get_update_result(self, update_id: str) -> Optional[UpdateResult]:
        """Get the result of a previously executed update."""
        with self._lock:
            return self._pending.get(update_id)

    def list_handlers(self) -> List[str]:
        """List registered update handler names."""
        return list(self._handlers.keys())

    def list_pending(self) -> List[str]:
        """List pending update IDs."""
        with self._lock:
            return list(self._pending.keys())

File: sdk/go/update/update.go
// Package update provides the Workflow Update API — synchronous workflow mutation.
//
// Unlike signals (fire-and-forget), updates provide:
//   - Synchronous request/response semantics
//   - Wait policies (Accepted, Completed, Admitted)
//   - Validation before execution
//   - Named update handlers registered by workflows
//
// Usage:
//
//	client := update.NewUpdateClient("localhost:7234")
//	client.RegisterHandler("setAmount", func(args interface{}) interface{} {
//	    return args
//	}, nil)
//	result, err := client.executeUpdate(context.Background(), update.UpdateRequest{
//	    WorkflowKey: 42,
//	    UpdateName:  "setAmount",
//	    Args:        map[string]int{"amount": 100},
//	})
package update

import (
	"fmt"
	"sync"
	"time"
)

// UpdateStatus represents the status of a workflow update.
type UpdateStatus int

const (
	Admitted  UpdateStatus = 0
	Accepted  UpdateStatus = 1
	Completed UpdateStatus = 2
	Rejected  UpdateStatus = 3
)

// UpdateWaitPolicy controls how long to wait for an update.
type UpdateWaitPolicy int

const (
	WaitAdmitted  UpdateWaitPolicy = 0
	WaitAccepted  UpdateWaitPolicy = 1
	WaitCompleted UpdateWaitPolicy = 2
)

// UpdateRequest represents a request to execute a workflow update.
type UpdateRequest struct {
	WorkflowKey uint64
	UpdateID    string
	UpdateName  string
	Args        interface{}
	WaitPolicy  UpdateWaitPolicy
}

// UpdateResult represents the result of a workflow update.
type UpdateResult struct {
	UpdateID   string
	Status     UpdateStatus
	Result     interface{}
	Error      string
	DurationMs float64
}

// UpdateHandler is a named handler for workflow updates.
type UpdateHandler struct {
	Name      string
	Handler   func(args interface{}) interface{}
	Validator func(args interface{}) bool
}

// UpdateClient provides workflow update operations.
type UpdateClient struct {
	serverAddress string
	handlers      map[string]*UpdateHandler
	pending       map[string]*UpdateResult
	mu            sync.RWMutex
}

// NewUpdateClient creates a new update client.
func NewUpdateClient(serverAddress string) *UpdateClient {
	return &UpdateClient{
		serverAddress: serverAddress,
		handlers:      make(map[string]*UpdateHandler),
		pending:       make(map[string]*UpdateResult),
	}
}

// RegisterHandler registers a named update handler.
func (c *UpdateClient) RegisterHandler(name string, handler func(args interface{}) interface{}, validator func(args interface{}) bool) {
	c.handlers[name] = &UpdateHandler{
		Name:      name,
		Handler:   handler,
		Validator: validator,
	}
}

// ExecuteUpdate executes a workflow update synchronously.
func (c *UpdateClient) ExecuteUpdate(req UpdateRequest) (*UpdateResult, error) {
	uid := req.UpdateID
	if uid == "" {
		uid = fmt.Sprintf("update-%d-%d", req.WorkflowKey, time.Now().UnixMilli())
	}
	start := time.Now()

	handler, ok := c.handlers[req.UpdateName]
	if !ok {
		result := &UpdateResult{
			UpdateID:   uid,
			Status:     Rejected,
			Error:      fmt.Sprintf("no handler registered for update '%s'", req.UpdateName),
			DurationMs: float64(time.Since(start).Milliseconds()),
		}
		c.mu.Lock()
		c.pending[uid] = result
		c.mu.Unlock()
		return result, nil
	}

	if handler.Validator != nil && !handler.Validator(req.Args) {
		result := &UpdateResult{
			UpdateID:   uid,
			Status:     Rejected,
			Error:      "update validation failed",
			DurationMs: float64(time.Since(start).Milliseconds()),
		}
		c.mu.Lock()
		c.pending[uid] = result
		c.mu.Unlock()
		return result, nil
	}

	value := handler.Handler(req.Args)
	result := &UpdateResult{
		UpdateID:   uid,
		Status:     Completed,
		Result:     value,
		DurationMs: float64(time.Since(start).Milliseconds()),
	}
	c.mu.Lock()
	c.pending[uid] = result
	c.mu.Unlock()
	return result, nil
}

// GetUpdateResult retrieves the result of a previously executed update.
func (c *UpdateClient) GetUpdateResult(updateID string) *UpdateResult {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.pending[updateID]
}

// ListHandlers returns the names of registered update handlers.
func (c *UpdateClient) ListHandlers() []string {
	names := make([]string, 0, len(c.handlers))
	for name := range c.handlers {
		names = append(names, name)
	}
	return names
}

// ListPending returns the IDs of pending updates.
func (c *UpdateClient) ListPending() []string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	ids := make([]string, 0, len(c.pending))
	for id := range c.pending {
		ids = append(ids, id)
	}
	return ids
}
